samples/msal-node-samples/client-credentials-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import axios from 'axios';
+import { performance } from "perf_hooks";
+import { RedisClientType } from "redis";
+import {
+    Configuration,
+    LogLevel,
+    ConfidentialClientApplication,
+    ClientCredentialRequest,
+    AuthenticationResult
+} from "@azure/msal-node";
+
+import CustomCachePlugin from "./CustomCachePlugin";
+import RedisClientWrapper from "./RedisClientWrapper";
+import AxiosHelper from './AxiosHelper';
+
+export type AppConfig = {
+    instance: string;
+    tenantId: string;
+    clientId: string;
+    clientSecret: string;
+};
+
+export class AuthProvider {
+    private msalConfig: Configuration;
+    private cacheClientWrapper: RedisClientWrapper;
+
+    private constructor(msalConfig: Configuration, cacheClient: RedisClientType) {
+        this.msalConfig = msalConfig;
+        this.cacheClientWrapper = new RedisClientWrapper(cacheClient);
+    }
+
+    static async initialize(appConfig: AppConfig, cacheClient: RedisClientType): Promise<AuthProvider> {
+        const msalConfig = {
+            auth: {
+                clientId: appConfig.clientId,
+                authority: `${appConfig.instance}/${appConfig.tenantId}`,
+                clientSecret: appConfig.clientSecret,
+            },
+            system: {
+                loggerOptions: {
+                    loggerCallback(loglevel, message, containsPii) {
+                        console.log(message);
+                    },
+                    logLevel: LogLevel.Trace,
+                    piiLoggingEnabled: false
+                },
+                // proxyUrl: "http://localhost:8888" // uncomment to capture traffic with Fiddler
+            }
+        } as Configuration;
+
+        const msalConfigWithMetadata = await AuthProvider.getMetadata(msalConfig, cacheClient);
+
+        return new AuthProvider(msalConfigWithMetadata, cacheClient);
+    }
+
+    async getToken(tokenRequest: ClientCredentialRequest): Promise<AuthenticationResult | null> {
+        const cca = new ConfidentialClientApplication({
+            ...this.msalConfig,
+            cache: {
+                cachePlugin: new CustomCachePlugin(
+                    this.cacheClientWrapper,
+                    `${this.msalConfig.auth.clientId}.${this.msalConfig.auth.authority!.split("/").pop()!}` // partitionKey <clientId>.<tenantId>
+                )
+            }
+        });
+
+        let tokenResponse = null;
+
+        try {
+            performance.mark("acquireTokenByClientCredential-start");
+            await cca.getTokenCache().getAllAccounts(); // required for triggering cache read
```

Comment:
Can we please log a bug or feature request for this? We should not have to do this.

Reply:
I did create a [bug report here](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/5502) a while ago. Would you like me to create one internally on ADO as well?